Initial addition of riscv-trace-spec signals to ibex_core (#949)
lowRISC/ibex: doc/riscv_trace.rst

.. _riscv_trace:                                                                                         

   :name: riscv-trace-spec

>    :align: center

Reviewer: Remove these two lines.

> RISC-V Trace Specs Interface

Reviewer: ```suggestion
RISC-V Trace Interface
```

============================

> Ibex implements a trace interface compatible with the riscv-trace-spec Version 1.0 issued in March 20, 2020. 

Reviewer: ```suggestion
Ibex implements an instruction trace interface compatible with the `RISC-V Trace Specification, version 1.0, as released on March 20, 2020 <https://github.com/riscv/riscv-trace-spec/>`_. 
```

> This interface currently exports needed internal signal as ibex_core outputs to be used to encode the trace packets to be logged.

Reviewer: Most of the time, you can remove words such as "currently" from the documentation. Everything can change, it is implied that the documentation refers to the current state of the implementation.

```suggestion
This interface exports needed internal signals as ibex_core outputs to be used to encode the trace packets to be logged.
```

> Signals implemented until now are shown in the table below:

Reviewer: You can remove this sentence.

Trace Signals
-------------

> For now, only single retirement per block is implemented. 

Reviewer: Ibex doesn't retire more than one instruction per cycle, and it's unlikely that this will ever change (famous last words). So you can remove this sentence.

                                                                       

+-----------------------+--------------------------------------+-------------------------------------------------------------+

> |   Signal Name         |            # of bits                 |                         Description                         |

Reviewer: To get the table to render properly and be more readable, you can make the columns more narrow. 

```suggestion
| Signal Name | # of bits | Description |
```

.. _riscv_trace:                                                                                         

RISC-V Trace Interface
============================

Ibex implements an instruction trace interface compatible with the `RISC-V Trace Specification, version 1.0, as released on March 20, 2020 <https://github.com/riscv/riscv-trace-spec/>`_. 
This interface exports needed internal signals as ibex_core outputs to be used to encode the trace packets to be logged.

Signals implemented until now are shown in the table below:

Trace Signals
-------------

+--------------------+---------------------+-------------------------------------------------------------+

> |   Signal Name      |   Number of bits    |                         Description                         |

Reviewer: Remove all unnecessary whitespace around these fields to make them less wide and the source more readable.

"Number of bits" can be shortened to "Bits"

+====================+=====================+=============================================================+
| rv_trace_ctype     |           2         | 0: Don’t report;                                            |
|                    |                     | 1: Report imprecisely;                                      |
|                    |                     | 2: Report precisely;                                        |

> |                    |                     | 3: Report as asynchronous discontinuity.                    |

Reviewer: We removed this signal.

+--------------------+---------------------+-------------------------------------------------------------+
| rv_trace_itype     |           3         | 0: Final instruction in block is none of other itype codes; |
|                    |                     | 1: Exception. An exception that traps occurred              |
|                    |                     | following the final retired instruction in the block;       |
|                    |                     | 2: Interrupt. An interrupt that traps occurred              |
|                    |                     | following the final retired instruction in the block;       |
|                    |                     | 3: Exception or interrupt return;                           |
|                    |                     | 4: Nontaken branch;                                         |
|                    |                     | 5: Taken branch;                                            |

> |                    |                     | 6: Uninferable jump if itype_width_p is 3                   |

Reviewer: Only document what we actually iplement:

```suggestion
|                    |                     | 6: Uninferable jump                 |
```

+--------------------+---------------------+-------------------------------------------------------------+

> | rv_trace_iaddr     |          32         | Contains the value of the address in the cases it needs     |

Reviewer: iaddr is *always* reported, there's no trace filtering happened yet.

Author: So you mean we will be selecting when to actually to report it later by the encoder, and here we just describe the interface in ibex-core?

Reviewer: Yes, the interface you're describing here is the core interface; it's effectively the text form of the signals you're creating in `ibex_core.sv`.

|                    |                     | to be reported in. Namely:                                  |
|                    |                     | -Jumps with target that can't be inferred from source code; |
|                    |                     | -The instruction retired immediately after a jump with a    |
|                    |                     | target that cannot be inferredfrom the source code (also    |
|                    |                     | referred to as the target or destination of the jump);      |
|                    |                     | -Taken and nontaken branches;                               |
|                    |                     | -Last instruction retired before an exception or interrupt; |
|                    |                     | -First instruction retired following an exception or intert;|
|                    |                     | -The last instruction retired before a privilege change;    |
|                    |                     | -The first instruction retired following a privilege change.|
+--------------------+---------------------+-------------------------------------------------------------+

> | rv_trace_cause     |          12         | ucause/scause/mcause CSR                                    |

Reviewer: There's no scause in Ibex.

+--------------------+---------------------+-------------------------------------------------------------+

> | rv_trace_tval      |          32         | Associated trap value, e.g. the faulting virtual address for|

Reviewer: - Ibex doesn't implement S mode, so there's no stval.
- Ibex doesn't support virtual memory, so instead of virtual address, it's just address.

|                    |                     | address exceptions, as written to the utval/stval/mtval CSR |
+--------------------+---------------------+-------------------------------------------------------------+

> | rv_trace_priv      |          2          | Privilege level for all instructions retired on this cycle. |

Reviewer: You can simplify this; Ibex never retires more than one instruction per cycle.

+--------------------+---------------------+-------------------------------------------------------------+

> | rv_trace_iretire   |          1          | Number of instructions retired in this block (0 or 1).      |                         

Reviewer: You can simplify this to say "The trace signals are valid (1), or not (0)."

+--------------------+---------------------+-------------------------------------------------------------+

Parameters
----------

> Referring to table 7.1, here are the parameters specified in the the implemented interface:

Reviewer: ```suggestion
Referring to Table 7.1 in the RISC-V Trace Specification v1.0, here are the parameters specified in the the implemented interface:
```
